src/libraries/Common/src/System/Reflection/Metadata/TypeNameParserHelpers.cs

```diff
@@ -0,0 +1,416 @@
+
+#if NET8_0_OR_GREATER
+            int offset = input.IndexOfAny(_endOfFullTypeNameDelimitersSearchValues);
+            if (offset < 0)
+            {
+                return input.Length; // no type name end chars were found, the whole input is the type name
+            }
+
+            if (input[offset] == EscapeCharacter) // this is very rare (IL Emit or pure IL)
+            {
+                offset = GetUnescapedOffset(input, startOffset: offset); // this is slower, but very rare so acceptable
+            }
+#else
+            int offset = GetUnescapedOffset(input, startOffset: 0);
+#endif
+            isNestedType = offset > 0 && offset < input.Length && input[offset] == '+';
+            return offset;
+
+            static int GetUnescapedOffset(ReadOnlySpan<char> input, int startOffset)
+            {
+                int offset = startOffset;
+                for (; offset < input.Length; offset++)
+                {
+                    char c = input[offset];
+                    if (c == EscapeCharacter)
+                    {
+                        offset++; // skip the escaped char
+
+                        if (offset == input.Length || // invalid name that ends with escape character
+                            !NeedsEscaping(input[offset])) // invalid name, escapes a char that does not need escaping
+                        {
+                            return -1;
+                        }
+                    }
+                    else if (NeedsEscaping(c))
+                    {
+                        break;
+                    }
+                }
+                return offset;
+            }
+
+            static bool NeedsEscaping(char c) => c is '[' or ']' or '&' or '*' or ',' or '+' or EscapeCharacter;
+        }
+
+        internal static ReadOnlySpan<char> GetName(ReadOnlySpan<char> fullName)
+        {
+            int offset = fullName.LastIndexOfAny('.', '+');
+
+            if (offset > 0 && fullName[offset - 1] == EscapeCharacter) // this should be very rare (IL Emit & pure IL)
+            {
+                offset = GetUnescapedOffset(fullName, startIndex: offset);
+            }
+
+            return offset < 0 ? fullName : fullName.Slice(offset + 1);
+
+            static int GetUnescapedOffset(ReadOnlySpan<char> fullName, int startIndex)
+            {
+                int offset = startIndex;
+                for (; offset >= 0; offset--)
+                {
+                    if (fullName[offset] is '.' or '+')
+                    {
+                        if (offset == 0 || fullName[offset - 1] != EscapeCharacter)
+                        {
+                            break;
+                        }
+                        offset--; // skip the escaping character
+                    }
+                }
+                return offset;
+            }
+        }
+
+        internal static string GetRankOrModifierStringRepresentation(int rankOrModifier, ValueStringBuilder builder)
```

Comment:
VSB passed byval - possible ArrayPool corruption.

Reply:
@GrabYourPitchforks even if this method calls `vsb.ToString()` at the end?

https://github.com/dotnet/runtime/blob/d63365fb7293b497a0bc7f17592c3d6439f1d451/src/libraries/Common/src/System/Reflection/Metadata/TypeNameParserHelpers.cs#L212

Which calls the `Dispose`?

https://github.com/dotnet/runtime/blob/e60d9a33355b536ba51478fc1940f1a4c8a799bc/src/libraries/Common/src/System/Text/ValueStringBuilder.cs#L90-L93

Reply:
Yes. Levi's point is this is taking a copy of the ValueStringBuilder. If it's wrapping an ArrayPool array, when you call ToString/Dispose, that array will be returned to the pool. But the caller of this method still has the original ValueStringBuilder, which may still be referencing that original array, won't see it as having been disposed of, and so if it also calls Dispose/ToString, that array will be doubly-returned. That could happen as well as part of just appending to it if it needs to return whatever array it had as a temporary.

Reply:
Thanks a lot for the explanation!